File: C++/Tree/boundaryTraversalOfBinaryTree.cpp
// try submitting this on GfG, here's the link for the problem:
// https://www.geeksforgeeks.org/problems/boundary-traversal-of-binary-tree/1

/*
class Node {
  public:
    int data;
    Node* left;
    Node* right;

    // Constructor to initialize a new node
    Node(int val) {
        data = val;
        left = NULL;
        right = NULL;
    }
};
*/

class Solution {
  public:
    vector<int> boundaryTraversal(Node *root) {
        // code here
        vector<int>ans;
        if(!isLeaf(root)) ans.push_back(root->data);
        addLeftBoundary(root->left, ans);
        addLeaves(root, ans);
        addRightBoundary(root->right, ans);
        return ans;    
    }
    void addLeftBoundary(Node* root, vector<int> &ans){
        if(root == NULL) return;
        while(root){
            if(!isLeaf(root)) 
                ans.push_back(root->data);
            if(root->left) root = root->left;
            else root=root->right;
            }
    }
    
    void addRightBoundary(Node* root, vector<int> &ans){
        if(root == NULL) return;
        vector<int>temp;
        while(root){
            if(!isLeaf(root)) 
                temp.push_back(root->data);
            if(root->right) root = root->right;
            else root=root->left;
        }
        for(int i  = temp.size()-1;i>=0;i--){
            ans.push_back(temp[i]);
        }
    }
    
    void addLeaves(Node* root, vector<int> &ans){
        if(root==NULL) return;
        
        if(isLeaf(root)) ans.push_back(root->data);
        addLeaves(root->left, ans);
        addLeaves(root->right, ans);
    }
    
    bool isLeaf(Node *root){
        if(root->left == NULL && root->right == NULL)
            return true;
        return false;
    }      
};

File: java/Tree/boundaryTraversalOfBinaryTree.java
//try submitting this on GFG, here's the link for the problem
//https://practice.geeksforgeeks.org/problems/boundary-traversal-of-binary-tree/1

/*
class Node {
    int data;
    Node left, right;

    public Node(int d) {
        data = d;
        left = right = null;
    }
}
*/

class Solution {
    ArrayList<Integer> boundaryTraversal(Node root) {
        // code here
        ArrayList<Integer> ans = new ArrayList<>();

        if (root == null) return ans;

        // Add root if it's not a leaf
        if (!isLeaf(root)) {
            ans.add(root.data);
        }

        // Left boundary
        addLeftBoundary(root.left, ans);

        // Leaves
        addLeaves(root, ans);

        // Right boundary
        addRightBoundary(root.right, ans);

        return ans;
    }

    private void addLeftBoundary(Node node, ArrayList<Integer> ans) {
        while (node != null) {
            if (!isLeaf(node)) {
                ans.add(node.data);
            }
            if (node.left != null) {
                node = node.left;
            } else {
                node = node.right;
            }
        }
    }

    private void addRightBoundary(Node node, ArrayList<Integer> ans) {
        ArrayList<Integer> temp = new ArrayList<>();

        while (node != null) {
            if (!isLeaf(node)) {
                temp.add(node.data);
            }
            if (node.right != null) {
                node = node.right;
            } else {
                node = node.left;
            }
        }

        // Add in reverse order
        for (int i = temp.size() - 1; i >= 0; i--) {
            ans.add(temp.get(i));
        }
    }

    private void addLeaves(Node node, ArrayList<Integer> ans) {
        if (node == null) return;

        if (isLeaf(node)) {
            ans.add(node.data);
            return;
        }

        addLeaves(node.left, ans);
        addLeaves(node.right, ans);
    }

    private boolean isLeaf(Node node) {
        return node.left == null && node.right == null;
    }
}